File: python_backend/app/application/use_cases/list_sessions.py
"""List user's sessions, paginated, with optional filters."""
from typing import Optional

from app.application.ports.sessions_repository import SessionsRepository
from app.domain.entities.session import Session


class ListSessionsUseCase:
    def __init__(self, repo: SessionsRepository) -> None:
        self.repo = repo

    def execute(
        self,
        *,
        user_id: str,
        scenario: Optional[str] = None,
        search: Optional[str] = None,
        limit: int = 20,
        offset: int = 0,
    ) -> list[Session]:
        # Clamp pagination to sane ranges.
        limit = max(1, min(limit, 100))
        offset = max(0, offset)
        return self.repo.list(
            user_id=user_id,
            scenario=scenario,
            search=search,
            limit=limit,
            offset=offset,
        )

File: web/src/application/use-cases/list-sessions.ts
/**
 * ListSessionsUseCase — GET /api/sessions with optional filters.
 *
 * Backend returns rows ordered by `started_at DESC`. Used by the
 * /app/sessions list page which needs scenario + search filtering and a
 * higher limit than the dashboard's 5 recent rows.
 */

import type { Session } from "@/domain/entities/session";
import type { SessionsApiPort } from "@/application/ports/sessions-api.port";

export interface ListSessionsParams {
  scenario?: string;
  search?: string;
  limit?: number;
  offset?: number;
}

export class ListSessionsUseCase {
  constructor(private readonly sessionsApi: SessionsApiPort) {}

  async execute(params: ListSessionsParams = {}): Promise<Session[]> {
    return this.sessionsApi.list(params);
  }
}
